server/api.go

```diff
@@ -272,3 +273,150 @@ func (d dexAPI) RevokeRefresh(ctx context.Context, req *api.RevokeRefreshReq) (*
 
 	return &api.RevokeRefreshResp{}, nil
 }
+
+// parseConnConfig will parse the provided connector config to ensure it is correct.
+func parseConnConfig(connType string, config []byte) error {
+	f, ok := ConnectorsConfig[connType]
+	if !ok {
+		return fmt.Errorf("unknown connector type %q", connType)
+	}
+
+	connConfig := f()
+	if len(config) != 0 {
+		if err := json.Unmarshal(config, connConfig); err != nil {
+			return fmt.Errorf("parse connector config: %v", err)
+		}
+	}
+	return nil
+}
+
+func (d dexAPI) CreateConnector(ctx context.Context, req *api.CreateConnectorReq) (*api.CreateConnectorResp, error) {
+	if req.Connector == nil {
+		return nil, errors.New("no connector supplied")
+	}
+	if req.Connector.Id == "" || req.Connector.Type == "" || req.Connector.Name == "" {
+		return nil, errors.New("Connector ID, Type, and Name are mandatory fields")
+	}
+	if len(req.Connector.Config) != 0 {
+		if err := parseConnConfig(req.Connector.Type, req.Connector.Config); err != nil {
+			return nil, fmt.Errorf("create connector: %v", err)
+		}
+	}
+
+	c := storage.Connector{
+		ID:              req.Connector.Id,
+		Type:            req.Connector.Type,
+		Name:            req.Connector.Name,
+		ResourceVersion: "1",
+		Config:          req.Connector.Config,
+	}
+	if err := d.s.CreateConnector(c); err != nil {
+		if err == storage.ErrAlreadyExists {
+			return &api.CreateConnectorResp{AlreadyExists: true}, nil
+		}
+		d.logger.Errorf("api: failed to create connector: %v", err)
+		return nil, fmt.Errorf("create connector: %v", err)
+	}
+
+	return &api.CreateConnectorResp{}, nil
+}
+
+func (d dexAPI) UpdateConnector(ctx context.Context, req *api.UpdateConnectorReq) (*api.UpdateConnectorResp, error) {
+	if req.Id == "" {
+		return nil, errors.New("no connector ID supplied")
+	}
+	if req.Type == "" && req.Name == "" && len(req.Config) == 0 {
+		return nil, errors.New("nothing to update")
+	}
+	if len(req.Config) != 0 {
```

Comment:
A user can update any one of these values req.Type, req.Name, req.Config or all of them. In the case that they just want to update the config we need to make the check again. It could be the case that they just want to update the type or the name.